File: C++/StackUsingQueue.cpp
#include <iostream>
#include <queue>

using namespace std;

class StackUsingQueues {
public:
    void push(int x) {
        tempQueue.push(x);
        while (!mainQueue.empty()) {
            tempQueue.push(mainQueue.front());
            mainQueue.pop();
        }

        // Swap the main and temporary queues (main = temp, temp = empty)
        mainQueue.swap(tempQueue);
    }

    void pop() {
        mainQueue.pop();
    }

    int top() {
        return mainQueue.front();
    }

    bool empty() {
        return mainQueue.empty();
    }

private:
    queue<int> mainQueue;
    queue<int> tempQueue;
};

int main() {

    StackUsingQueues stack;

    stack.push(1);
    stack.push(2);
    stack.push(3);
    stack.push(4);
    stack.push(5);

    cout << "Top element: " << stack.top() << endl; // Output: 5

    stack.pop();

    cout << "Top element after pop: " << stack.top() << endl; // Output: 4

    stack.push(6);

    cout << "Top element: " << stack.top() << endl; // Output: 6

    system("pause");
    return 0;
}

File: JAVA/StackUsingQueue.java
import java.util.LinkedList;
import java.util.Queue;

public class StackUsingQueue {
    private Queue<Integer> queue1;
    private Queue<Integer> queue2;

    public StackUsingQueue() {
        queue1 = new LinkedList<>();
        queue2 = new LinkedList<>();
    }

    public void push(int element) {
        queue2.offer(element);
        while (!queue1.isEmpty()) {
            queue2.offer(queue1.poll());
        }
        Queue<Integer> temp = queue1;
        queue1 = queue2;
        queue2 = temp;
    }

    public int pop() {
        if (queue1.isEmpty()) {
            return -1; // Stack is empty
        }
        return queue1.poll();
    }

    public int top() {
        if (queue1.isEmpty()) {
            return -1; // Stack is empty
        }
        return queue1.peek();
    }

    public boolean isEmpty() {
        return queue1.isEmpty() && queue2.isEmpty();
    }
}
